### ISSUE TITLE
Haddock-2.18 hangs forever (ghc-8.2) when using makeClassy

### ISSUE THREAD

**Issue by PierreR** [haskell/haddock#679]
`haddoc-2.18.1`  (ghc-8.2.1) will hang forever with `stack haddock` or `cabal haddock` for the `language-puppet`  package (v-1.3.10)

With the same source, both command works flawlessly in `ghc-8.0.2` (`haddock-2.17.3`)

I have been trying to fix this with `{-# OPTIONS_HADDOCK hide, prune  #-}` without success. `haddock` would hang on another module and spicing modules with the pragma leads nowhere.

There are no error message. The `cpu` level just goes off and the process hangs forever. 

Is there any sane way to debug this ? The fact that it just hangs is so abnormal.

See https://github.com/fpco/stackage/issues/2822

**Comment by alexbiehl** [comment]
I will investigate. But it sounds like a GHC thing to me. Hoping to find
out why.

Pi3r <notifications@github.com> schrieb am Mo., 4. Sep. 2017 um 13:49 Uhr:

**Comment by PierreR** [comment]
@alexbiehl Thanks. Here is the [travis build](https://travis-ci.org/bartavelle/language-puppet) 

I am also able to reproduce this locally on my machine (nixos)

**Comment by PierreR** [comment]
In case it helps this is the end output with `strace`:
```
    puppetDBFacts (src/Facter.hs:174)
  93% ( 27 / 29) in 'Puppet.Interpreter.Resolve'
) = ? ERESTARTSYS (To be restarted if SA_RESTART is set)
--- SIGVTALRM {si_signo=SIGVTALRM, si_code=SI_TIMER, si_timerid=0, si_overrun=0, si_value={int=0, ptr=NULL}} ---
rt_sigreturn({mask=[]})                 = 202
futex(0x21b8254, FUTEX_WAIT_PRIVATE, 0, NULL) = ? ERESTARTSYS (To be restarted if SA_RESTART is set)
--- SIGVTALRM {si_signo=SIGVTALRM, si_code=SI_TIMER, si_timerid=0, si_overrun=0, si_value={int=0, ptr=NULL}} ---
rt_sigreturn({mask=[]})                 = 202
futex(0x21b8254, FUTEX_WAIT_PRIVATE, 0, NULL) = ? ERESTARTSYS (To be restarted if SA_RESTART is set)
--- SIGVTALRM {si_signo=SIGVTALRM, si_code=SI_TIMER, si_timerid=0, si_overrun=0, si_value={int=0, ptr=NULL}} ---
rt_sigreturn({mask=[]})                 = 202
```

**Comment by PierreR** [comment]
I have managed to pinpoint the problem. It is actually related to `makeClassy`. For instance if I replace `makeClassy` with `makeLenses` here, it is going to pass (after removing all other module deps of course):

https://github.com/PierreR/language-puppet/blob/cosmetic/src/Hiera/Core.hs#L82

To fix this I actually need to replace all `makeClassy` in the code base with `makeLenses` ...

**Comment by alexbiehl** [comment]
Fixed by #723
